fiftyone/utils/multiprocessing.py

```diff
@@ -0,0 +1,376 @@
+    if shard_method == "slice":
+        n = len(sample_collection)
+    else:
+        ids = sample_collection.values("id")
+        n = len(ids)
+
+    if num_workers is None:
+        num_workers = fou.recommend_process_pool_workers()
+
+    # Must cap size of select(ids) stages
+    if shard_method == "id":
+        max_shard_size = fou.recommend_batch_size_for_value(
+            ObjectId(), max_size=100000
+        )
+
+        if shard_size is not None:
+            shard_size = min(shard_size, max_shard_size)
+        elif n > num_workers * max_shard_size:
+            shard_size = max_shard_size
+
+    if shard_size is not None:
+        # Fixed size shards
+        edges = list(range(0, n + 1, shard_size))
+        if edges[-1] < n:
+            edges.append(n)
+    else:
+        # Split collection into exactly `num_workers` shards
+        edges = [int(round(b)) for b in np.linspace(0, n, num_workers + 1)]
+
+    if shard_method == "slice":
+        # Slice batches
+        slices = list(zip(edges[:-1], edges[1:]))
+    else:
+        # ID batches
+        slices = [ids[i:j] for i, j in zip(edges[:-1], edges[1:])]
+
+    num_shards = len(slices)
+    batches = list(
+        zip(
+            range(num_shards),
+            itertools.repeat(num_shards),
+            slices,
+        )
+    )
+
+    num_workers = min(num_workers, num_shards)
+
+    return batches, num_workers
+
+
+def _init_worker(dataset_name, view_stages, m, s, r, p, l):
+    from tqdm.auto import tqdm
+
+    import fiftyone as fo
+    import fiftyone.core.odm.database as food
+    import fiftyone.core.view as fov
+
+    global sample_collection, map_fcn, save, return_outputs, progress
+
+    # Ensure that each process creates its own MongoDB clients
+    # https://pymongo.readthedocs.io/en/stable/faq.html#using-pymongo-with-multiprocessing
+    food._disconnect()
+
+    dataset = fo.load_dataset(dataset_name)
+    if view_stages:
+        sample_collection = fov.DatasetView._build(dataset, view_stages)
+    else:
+        sample_collection = dataset
+
+    map_fcn = m
+    save = s
+    return_outputs = r
+    progress = p
+
+    if l is not None:
+        tqdm.set_lock(l)
+
+
+def _map_batch(input):
+    i, num_batches, batch = input
+
+    if isinstance(batch, tuple):
+        # Slice batches
+        start, stop = batch
+        total = stop - start
+        batch_view = sample_collection[start:stop]
+    else:
+        # ID batches
+        sample_ids = batch
+        total = len(sample_ids)
+        batch_view = fov.make_optimized_select_view(
+            sample_collection, sample_ids
+        )
+
+    if return_outputs:
+        values = {}
+
+    if progress:
+        desc = f"Batch {i + 1:0{len(str(num_batches))}}/{num_batches}"
+        with tqdm(total=total, desc=desc, position=i) as pb:
+            for sample in batch_view.iter_samples(autosave=save):
+                result = map_fcn(sample)
+                if return_outputs and result is not None:
```

Comment:
Something we can explicitly call out in docstring, in the aggregate/reduce step we will filter None return value